### ISSUE TITLE
docker internal dns use only the first nameserver declare in resolv.conf

### ISSUE THREAD

**Issue by JrCs** [moby/moby#20494]
Hi,

i have a server with 2 nameservers declare in /etc/resolv.conf.
The first one is the dns from consul to resolve nodes and services and the second is the nameserver to resolve internet adresses.

i create a new network bridge with the command:

``` bash
docker network create newnetwork
```

and attach a new container  with it

``` bash
docker run -it --rm --net=newnetwork busybox sh
```

In that container i can resolve nodes and services return by the dns of consul, but i can't resolv internet adresses.
So i think the internal dns of docker use only the first nameserver declare in resolv.conf instead of using all.

My configuration:

`docker version:`

``` pre
Client:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.4.3
 Git commit:   e21da33
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.4.3
 Git commit:   e21da33
 Built:
 OS/Arch:      linux/amd64
```

`docker info:`

``` pre
Containers: 72
 Running: 64
 Paused: 0
 Stopped: 8
Images: 75
Server Version: 1.10.0
Storage Driver: btrfs
 Build Version: Btrfs v4.2.2
 Library Version: 101
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.1-coreos
Operating System: CoreOS 955.0.0
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.26 GiB
Name: sd-65534
ID: 6F5O:FVKR:FMX3:B6KD:WCIF:3KLF:MYK2:65R5:OEHA:4GZT:RZKF:JTPK
Username: jrcs
Registry: https://index.docker.io/v1/
```

`uname -a:`

``` pre
Linux sd-65534 4.4.1-coreos #2 SMP Thu Feb 11 01:49:47 UTC 2016 x86_64 Intel(R) Xeon(R) CPU E3-1230 v3 @ 3.30GHz GenuineIntel GNU/Linux
```

**Comment by phemmer** [comment]
That's not how `resolv.conf` is supposed to be used. From `man 5 resolv.conf`:

> The algorithm used is to try a name server, and if the query times out, try the next, until out of name servers, then repeat trying all the name servers until a maximum number of retries are made.

Meaning all of the name servers listed should return the same results. If the first name server returns a NXDOMAIN error (not found), that's not a timeout, and the resolver is not required to proceed to the next nameserver.

You probably want to be using something like [dnsmasq](http://www.thekelleys.org.uk/dnsmasq/doc.html), which can be used for routing queries for different zones to different nameservers.

**Comment by cttttt** [comment]
Hmm...we're having a similar issue.  After a bit of investigation, the first DNS server in our list seems to have been returning `SERVFAIL`.

Tools like `nslookup`, `ping`, and others move onto the next server when this occurs.  For some reason, `docker` doesn't, returning a similar error to the one originally reported.

@phemmer should `docker` be moving on to the next server on `SERVFAIL`?

**Comment by phemmer** [comment]
Yes, docker should ("should" as in docker should be coded to do so if it is no, not that docker is coded and it should be working).

**Comment by sanimej** [comment]
@cttttt Yes, currently the embedded server is not checking specifically for `SERVFAIL` and won't try the other servers. This can be fixed.

But if the reason for `SERVFAIL` is because of the domain's authoritative name servers not responding then trying the next resolver in the list will not help. You might want to check the reason why are you are getting `SERVFAIL`

**Comment by djsly** [comment]
We are also facing this issues with the kube2sky image from kubernetes. We have two name server where the first one returns SERVFAIL, outside of Docker this seem to work fine with the line
````
;; Got SERVFAIL reply from xx.yy.zz.ww, trying next server
````

in Docker, it stops at the first nameserver

**Comment by marcosnils** [comment]
libnetwork related issue https://github.com/docker/libnetwork/issues/1664

**Comment by UBNT-Petr** [comment]
Some update on this issue? Some ETA?
We are facing the same problem.

**Comment by tomthetommy** [comment]
We're having the same issue here;

#/etc/resolv.conf
search kittens.org
nameserver 127.0.0.11
options ndots:0

First server responds with SERVFAIL, but then doesn't bother to try next one.

This works fine on the host using ping and nslookup, the next server is then used. Within the containers this doesn't happen.

The containers DO move onto another server if there is a timeout, other than that, it doesn't.

**Comment by Sytten** [comment]
The current workaround is to specify the --dns to the containers who need to resolve names.
If you take a look at the libnetwork issue, it has been fixed but not released.

**Comment by thaJeztah** [comment]
The fix for `ServFail` (https://github.com/docker/libnetwork/issues/1664, PR: https://github.com/docker/libnetwork/pull/2121), is  included in Docker 18.03.1; https://github.com/docker/docker-ce/blob/v18.03.1-ce/components/engine/vendor/github.com/docker/libnetwork/resolver.go#L494-L498

But it looks like there's other situations where a DNS returns no response, or no `A` / `AAAA` records; not sure if that's a valid reason to continue with the next DNS (see https://github.com/docker/compose/issues/5991#issuecomment-392758621); I'll open a libnetwork PR to discuss that

**Comment by hetykai** [comment]
When i run container with "--network yourCustomNet", that would raise the issue.

`docker run -it --rm --network yourCustomNet --dns-opt=rotate --dns-opt="timeout:1" --dns 202.112.20.131 --dns 8.8.8.8 alpine`
Run command `ping baidu.com` return `ping: bad address 'baidu.com'`
Check resolv.conf `cat /etc/resolv.conf`, this file does not contain the DNS IP address specified above.

Then i run `echo -e 'nameserver 202.112.20.131\nnameserver 8.8.8.8' >> /etc/resolv.conf`
And i check resolv.con `cat /etc/resolv.conf`, it contains the DNS IP address specified above, and i `ping baidu.com` return right ip.

When i run container without "--network yourCustomNet", it does NOT raise the issue, and resolv.conf contains the DNS IP address specified above.

So i think this issue cause by that the resolv.conf file could not work correctly with custom network.

![powershell_2018-08-24_20-51-25](https://user-images.githubusercontent.com/3378432/44586052-f7ff0b80-a7e0-11e8-9a96-5fcfe59e861e.png)

**Comment by hetykai** [comment]
My version is 18.03.1-ce-win65 (17513)

**Comment by thaJeztah** [comment]
I think that this should be resolved with these PR's merged;

- https://github.com/docker/libnetwork/pull/2121 Retry other external DNS servers on ServFail
- https://github.com/docker/libnetwork/pull/2171 Handle NXDOMAIN, REFUSED and log errors
  - vendored in https://github.com/moby/moby/pull/38031

**Comment by erdnaxeli** [comment]
It looks like the issue is happening again with docker 19.03.9.

How to reproduce:
1. `docker network create test`
2. `docker run --rm --net=test busybox nslookup perdu.com` => success
3. edit /etc/resolv.conf and add a non working nameserver entry at the top
4. `docker run --rm --net=test busybox nslookup perdu.com` => `;; connection timed out; no servers could be reached`

docker version:
```
Client: Docker Engine - Community
 Version:           19.03.9
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        9d988398e7
 Built:             Fri May 15 00:24:16 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.9
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       9d988398e7
  Built:            Fri May 15 00:22:48 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

docker info:
```
Client:
 Debug Mode: false

Server:
 Containers: 16
  Running: 9
  Paused: 0
  Stopped: 7
 Images: 22
 Server Version: 19.03.9
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.9.0-8-amd64
 Operating System: Debian GNU/Linux 9 (stretch)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.58GiB
 Name: ns3033797.ip-91-121-80.eu
 ID: CSRI:7UFR:M4GX:3T2L:EPTE:VQO7:Z4TS:MJ4R:NRMM:6W6C:UOUP:5VP5
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by Sytten** [comment]
There was a big update to the DNS client provided by the vendor [miekg](https://github.com/miekg/dns) here:
https://github.com/moby/libnetwork/commit/d5e341e6798c619147691d53ceb6b426b3b8cb9d#diff-6756fdfda338c13b910a6416f2ac9f54

This is most likely the root of the issue.

**Comment by thaJeztah** [comment]
@erdnaxeli your case looks like a duplicate of https://github.com/moby/moby/issues/41003, which will be addressed by https://github.com/moby/libnetwork/pull/2551 (once merged into the 19.03 branch and vendored in this repository)

**Comment by thaJeztah** [comment]
I think the other / original issues reported in this ticket have been addressed by fixes that have already been released (see https://github.com/moby/moby/issues/20494#issuecomment-433124382), so I'm closing this ticket.

For the 19.03.9 issue, subscribe to https://github.com/moby/moby/issues/41003 to stay informed on updates

**Comment by erdnaxeli** [comment]
> @erdnaxeli your case looks like a duplicate of #41003, which will be addressed by [moby/libnetwork#2551](https://github.com/moby/libnetwork/pull/2551) (once merged into the 19.03 branch and vendored in this repository)

It is indeed, even if actually its #41003 which is a duplicate of my issue :D

Good to see there is a fix.

**Comment by thaJeztah** [comment]
Your issue was reported as comment on an existing issue, so I prefer using https://github.com/moby/moby/issues/41003, to prevent multiple issues being conflated (which makes it confusing for readers "which" issue is being talked about).
